Fixes #33435 - Other Content Types Landing Page UI

rverdile opened the pull request:
This adds a new tab to the navigation bar: `Content Types > Other Content Types`. This links to a listing of content types, always filtered by one type. The only type supported here is Python Packages, which is selected by default.

Navigating to this page, once you have python packages created, you'll see a table with  "Name" and "Version" headers and those packages listed. It's also searchable with autocomplete. If not content has been created, there's a message saying so.

It should be possible to add more content types to this page by adding them to `ContentConfig.js`, in a similar structure to what is already there for python packages. That might be a fun test.

rverdile commented:
Not sure how to resolve the remaining eslint issues.

rverdile commented:
this should merge first: https://github.com/Katello/katello/pull/9629

sjha4 commented:
Empty page with `Uncaught Error: Too many re-renders. React limits the number of renders to prevent an infinite loop.` on loading page with no python packages..

sjha4 commented on webpack/scenes/Content/ContentActions.js:
```diff
@@ -0,0 +1,17 @@
+import api from '../../services/api';
+import { CONTENT_KEY, CONTENT_TYPES_KEY } from './ContentConstants';
+
+export const getContent = contentType => get({
+  type: API_OPERATIONS.GET,
+  key: CONTENT_KEY,
+  url: api.getApiUrl(`/${contentType}`),
+});
```
```suggestion
export const getContent = (contentType, params) => get({
  type: API_OPERATIONS.GET,
  key: CONTENT_KEY,
  params,
  url: api.getApiUrl(`/${contentType}`),
});
```

This will make pagination work..The pagination props are passed as params to the API.

sjha4 commented on webpack/scenes/Content/Table/ContentTable.js:
```diff
@@ -0,0 +1,117 @@
+    }
+  }, [enabledTypesResponse, enabledTypesStatus]);
+
+  const fetchItems = useCallback(
+    () => getContent(contentTypes[contentTypeSelected][1]),
+    [contentTypes, contentTypeSelected],
+  );
```
```suggestion
  const fetchItems = useCallback(
    (params) => getContent(contentTypes[contentTypeSelected][1], params),
    [contentTypes, contentTypeSelected],
  );
```
Passing params to fetchItem will make pagination and search work..

sjha4 commented on webpack/scenes/Content/Table/ContentTable.js:
```diff
@@ -0,0 +1,117 @@
+  function buildRows(details) {
+    const rows = [];
+    if (details) {
+      columnHeaders.forEach(header =>
+        rows.push(<Td>{header.getProperty(details)}</Td>));
```
```suggestion
      columnHeaders.forEach((header, index) =>
        rows.push(<Td key={index}>{header.getProperty(details)}</Td>));
```
This will resolve the console error about needing unique keys. You will need to disable an eslint rule to use this.
`/* eslint-disable react/no-array-index-key */`

rverdile commented:
updated. thanks @sjha4

sjha4 commented on webpack/scenes/Content/Table/ContentTable.js:
```diff
@@ -0,0 +1,118 @@
+
+  useEffect(() => {
+    dispatch(getContentTypes());
+  }, []); // eslint-disable-line react-hooks/exhaustive-deps
```
You can add dispatch as dependency here and remove the eslint comment..

sjha4 commented on webpack/scenes/Content/Table/ContentTable.js:
```diff
@@ -0,0 +1,118 @@
+  const response = useSelector(selectContent);
+  const enabledTypesResponse = useSelector(selectContentTypes);
+  const enabledTypesStatus = useSelector(selectContentTypesStatus);
+  const [contentTypeSelected, setContentTypeSelected] = useState('Python Packages');
```
You could move this to the parent ContentPage and pass down as a prop..This will let the Header on Page change with type selection.

rverdile replied:
cool!

sjha4 commented on webpack/scenes/Content/Table/ContentTable.js:
```diff
@@ -0,0 +1,118 @@
+      variant={TableVariant.compact}
+      autocompleteEndpoint={`/${contentTypes[contentTypeSelected][1]}/auto_complete_search`}
+      emptyContentTitle={`You currently don't have any ${contentTypeSelected}.`}
+      emptyContentBody={`${contentTypeSelected} will appear here when created.`}
```
if you add `additionalListeners={[contentTypeSelected]}` to the Table here, it will update when content type is picked in the selector. That will be needed when we use this page for other content types..

sjha4 commented on webpack/scenes/Content/Table/ContentTable.js:
```diff
@@ -0,0 +1,118 @@
+      <Tbody>
+        {results?.map(details => (
+          <Tr key={`${details.pulp_id}`}>
+            {buildRows(details)}
```
Not all of our content rabls return a pulp_id but I believe all return an id which will also be needed to route to details page. Maybe use details.id instead of pulp_id.

jlsherrill commented on webpack/scenes/Content/ContentConfig.js:
```diff
@@ -0,0 +1,11 @@
+
+export default () => [
+  {
+    names: { title: __('Python Packages'), plural: 'python_packages', singular: 'python_package' },
```
is there any way we can have this information pulled from the api?

rverdile replied:
Just the names or all of the information? Having the separate config file for this might be necessary to allow different content types to render information with different react components.

jlsherrill replied:
just the names, but potentially the columns in the future as well.  @sjha4 do you have any ideas how we could do this?  the idea is still that we can drop in the repository_type definition on the ruby server and thats it, the UI will adapt as needed

rverdile replied:
I don't know how much variation in the react components is expected, but maybe when a react component is desired it can just be associated with a keyword from the api.

sjha4 replied:
I like this approach of having a UI config file to control what shows on the UI..

But if we want to make it more generic, we could do a default list of columns to display (name, version) to begin with and get this config file to loop over content types and form this config automatically. We could still override a content config if we wanted to.

jlsherrill replied:
I'm not necessarily opposed to it, it just increases the footprint of where things need to be added.  If we do go this route, we'll need something that disables content types that aren't actually enabled on the server (for example, we wouldn't want to show python packages if python isn't enabled)

rverdile replied:
What if I add a kind of default that can have column headers added to it from the registration file, this way the only time it would need to be reconfigured on the javascript side is if different react components need to be specified.

And I'll make the default selected type whatever the first type returned from the API is, rather than python packages specifically.

sjha4 replied:
> And I'll make the default selected type whatever the first type returned from the API is, rather than python packages specifically.

This will probably need to be plucked out of the URL in the future..
For example: If user is going to ostree nav link, url changes to /ostree and that'd be the selection in the dropdown..

rverdile replied:
In that case, when someone clicks "Other Content Types" in the navbar, should it go to `/content/:content_type` or `/:content_type`, rather than just `/content`?

sjha4 replied:
/content is fine for now..We will need to do some changes when we route explicit content_type types directly to this page.. For ex: routing /ansible_collection would route to this page with contentTypeSelected reading from the URL..

rverdile commented:
Updated. Added test/fixtures and @sjha4's suggestions.

rverdile commented:
I added a default config that uses information from the registration file. It will let you add an arbitrary number of column headers to the table. You can also override the default config by adding the configuration to `ContentConfig.js` instead. 

The `names` hash still needs to be specified in the registration file. An example of what an override config looks like is in a comment in `ContentConfig.js`

I also moved the API logic from `ContentTable.js` to `ContentPage.js`

sjha4 commented on webpack/scenes/Content/ContentConfig.js:
```diff
@@ -0,0 +1,62 @@
+      types.push(key);
+    }
+  });
+  return types;
```
What is this doing exactly?

rverdile replied:
It's checking the enabled content types against what has a configuration defined in `ContentConfig()`. It returns a list of keys for content types that are enabled _and_ have a configuration specified. If a content type is in this list, it won't use `DefaultConfig()`.

It might benefit from a different name.

sjha4 replied:
I wasn't able to get it working with an override config locally..Can you verify that it works for you and let me know the config you used?

sjha4 commented on app/services/katello/repository_type.rb:
```diff
@@ -154,11 +160,19 @@ def initialize(options)
         self.content_type = options[:content_type]
         self.filename_key = options[:filename_key]
         self.duplicates_allowed = options[:duplicates_allowed]
+        self.ui_config = options[:ui_config]
```
Can we move this to parent ContentType so it's available for non-generic types too..You could add ui_config to to_json of parent and remove the overriding that limits it to generic types..Also would super() be helpful here for the common bits of initialize..

sjha4 commented on webpack/scenes/Content/ContentPage.js:
```diff
@@ -0,0 +1,73 @@
+  const enabledTypesResponse = useSelector(selectContentTypes);
+  const enabledTypesStatus = useSelector(selectContentTypesStatus);
+  const [contentTypeSelected, setContentTypeSelected] = useState(null);
+  const [typeSelectedConfig, setTypeSelectedConfig] = useState(null);
```
nitpick: Thoughts on renaming these to selectedContentType and selectedContentTypeConfig?

rverdile replied:
I agree :)

Feel free to comment on the names of other things too. Sometimes naming is hard.

jeremylenz commented on webpack/scenes/Content/Table/ContentTable.js:
```diff
@@ -0,0 +1,102 @@
+        </Tr>
+      </Thead>
+      <Tbody>
+        {results?.map(details => (
```
:heart_eyes:

rverdile commented:
Reverted most of the previous changes. API logic is still in `ContentPage.js` to check that the list of selectable content types contains only enabled content types.

rverdile commented:
Added message to display when no content types are found.

sjha4 commented on webpack/scenes/Content/ContentPage.js:
```diff
@@ -0,0 +1,73 @@
+  const [selectedContentType, setSelectedContentType] = useState(null);
+  const [contentTypes, setContentTypes] = useState(null);
+
+  function buildContentTypes(response) {
```
```suggestion
  const buildContentTypes = () => {
```
The param shouldn't need to be passed here. Can directly use the contentTypesResponse

Andrewgdewar replied:
also! this is es5 syntax, we tend to want to avoid that in our javascript react code now since about 2012. 
Here's an in depth breakdown of what I mean: 
[es5 vs es6 functions (the big jump)](https://engineering.carsguide.com.au/es5-vs-es6-syntax-6c8350fa6998)
[all the more modern methods](https://www.freecodecamp.org/news/es5-to-esnext-heres-every-feature-added-to-javascript-since-2015-d0c255e13c6e/)

sjha4 commented on webpack/scenes/Content/ContentPage.js:
```diff
@@ -0,0 +1,73 @@
+    dispatch(getContentTypes());
+  }, [dispatch]);
+
+  if (!contentTypes) {
```
```suggestion
  if (contentTypesStatus === STATUS.LOADING) {
```

jlsherrill reviewed (approved):
ACK!  Good work @rverdile !

sjha4 commented on webpack/scenes/Content/Table/ContentTable.js:
```diff
@@ -0,0 +1,90 @@
+        error,
+        status,
+      }}
+      variant={TableVariant.compact}
```
```suggestion
      key={contentTypes[selectedContentType][1]}
      variant={TableVariant.compact}
```

So I noticed the autosearch params are not updated when the selectedType changes. This will fix that.

jeremylenz replied:
Changing the key will cause the entire TableWrapper to unmount and remount.  It would work but seems a bit drastic.  I'm confused because I would think adding `selectedContentType` to `additionalListeners` should be all you need?

sjha4 replied:
So the search component inside tableWrapper is not refreshed with the additionalListener..That causes the autosearch Endpoint to not change. When switching between different content types we need to update the search bar with correct endpoints..Hence the complete refresh of the TableWrapper.

sjha4 replied:
We should actually remove the selectedContentType from additionalListeners since it's not needed anymore given we are re-rendering the whole table instead.

rverdile commented:
updated

rverdile commented on webpack/scenes/Content/ContentPage.js:
```diff
@@ -0,0 +1,73 @@
+
+  useEffect(() => {
+    // Set list of enabled content types and initial content type
+    const buildContentTypes = () => {
```
I moved `buildContentTypes` inside of the `useEffect` because it became a dependency after I changed the function syntax. Not sure if this is the best way to handle it.

sjha4 replied:
Ya..That's good..

sjha4 commented:
[test katello]

rverdile commented:
removed additionalListeners prop in `ContentTable` since it's being re-rendered with key

sjha4 reviewed (approved):
Great work @rverdile !! Works great and code looks good. 👍🏽 🥳
